lib/chef/client.rb

```diff
@@ -593,6 +613,39 @@ def save_updated_node
       end
     end
 
+    def setup_train
+      require "train"
+      protocol = Chef::Config.target_mode.protocol
+      train_config = Chef::Config.target_mode.to_hash.select { |k| Train.options(protocol).key?(k) }
+      train_config[:logger] = logger
+
+      # Train handles connection retries for us
+      @train_connection = Train.create(protocol, train_config).connection
+      train_connection.wait_until_ready
+    rescue SocketError => e # likely a dns failure, not caught by train
+      e.message.replace "Error connecting to #{train_connection.uri} - #{e.message}"
+      raise e
+    rescue Train::PluginLoadError
+      logger.error("Invalid target mode protocol: #{protocol}")
+      exit(false)
+    end
```

Comment:
sort of wonder here about the inherent tight coupling here between one train connection and one client connection.

particularly when we start talking next-gen chef-provisioning-like ideas, we may want to have this more loosely coupled.

i'm leaning towards run_context.train_connection should be the primary object, you can have a delegator here for train_connection -> run_context.train_connection.

then access to train_connection should probably have a lazy initializer so that when it is hit the first time it calls setup_train in a method off of the run_context.

you should probably have a chef class like Chef::TrainThing which wraps a single train connection which is the thing which actually hangs off of the run_context, then the run_context just calls Chef::TrainThingy.build_connection() to get a new object -- then the run_context and the client go back to "knowing" very little about train.

Reply:
We do need a train connection before the run context is created because we use it to go get ohai-like-data at the same time as ohai would, but I suppose there's no reason to keep that particular train connection for later. some day we might have ohai go out and make that connection instead, so could split that now too.

Reply:
I think I'm going to change the run_context in Chef 15 to get created way the hell early, and decouple the creation of the run_context from the compilation of the cookbooks, so it just becomes the global-ish state context of the whole run.  keep running into no way of threading state through the actual whole run being a problem.